2 feature 2 using items

Grace199 opened the pull request:
Grace - Added Item and Inventory Classes. Added UseItemRunner use case along with it's Input and Output Boundaries. Also included the dummy presenter. Ignore the sortedByQuantity() method in the Inventory class.

Grace199 committed:
Grace - Built Item and Inventory Classes. Built UseItemRunner and the Input and Output Boundaries. Also included a dummie presenter for the UseItemRunner.

Grace199 committed:
Grace - Built Item and Inventory Classes. Built UseItemRunner and the Input and Output Boundaries. Also included a dummie presenter for the UseItemRunner.

Samuel-Glasroth reviewed (dismissed):
In the removeItem function in the Inventory class you have a variable using snake case instead of camel case
In the addItem function in the Inventory class you have some redundant code which can be fixed by separating your if statements
Otherwise looks good

Grace199 committed:
Grace - Item and Inventory now implement Serializable.

Grace199 committed:
Grace - Item and Inventory now implement Serializable
      - Removed redundant code in the addItem() method in the Inventory class
      - Added the AddItem use case (with it’s interactor, input and output boundary interfaces, and input and output data structures)
      - Added the RemoveItem use case (with it’s interactor, input and output boundary interfaces, and input and output data structures)
      - Added more documentation for the UseItem use case

adimaggio2021 commented on src/main/java/Mariogatchi/AddItemRequestModel.java:
```diff
@@ -0,0 +1,35 @@
+package Mariogatchi;
+
+public class AddItemRequestModel {
+    private Item item;
```
It would be helpful to put comments on each instance variable so we know what each is for.

Grace199 replied:
comments added

adimaggio2021 commented on src/main/java/Mariogatchi/Inventory.java:
```diff
@@ -1,4 +1,89 @@
+    public boolean removeItem(Item item, int quantityToRemove) {
+        if (itemToQuantity.containsKey(item) && quantityToRemove <= itemToQuantity.get(item)){
+            int newQuantity = itemToQuantity.get(item) - quantityToRemove;
+            itemToQuantity.replace(item, newQuantity);
```
Will this leave ghost items in the inventory if the newQuantity is zero?

Grace199 replied:
It will now remove the item if the new quantity is 0

adimaggio2021 commented on src/main/java/Mariogatchi/Item.java:
```diff
@@ -0,0 +1,67 @@
+    private Set<Environment> eligibleAreas = new Set<Environment>;
+
+    // a mapping of Effects to their corresponding values
+    private Map<Effect, Integer> effectValues = new Map<Effect, Integer>;
```
This can probably be implemented as String to Integer.

Grace199 replied:
I made Effect an enum so is now a String constant.

adimaggio2021 commented on src/main/java/Mariogatchi/RemoveItemRunner.java:
```diff
@@ -0,0 +1,31 @@
+     */
+    @Override
+    public RemoveItemResponseModel removeItemFromInv(RemoveItemRequestModel requestModel) {
+        if (requestModel.getInventory().removeItem(requestModel.getItem(), requestModel.getQuantity())) {
```
Should you be able to remove Reusable items from your inventory. Could break some game functionality.

adimaggio2021 commented on src/main/java/Mariogatchi/UseItemRunner.java:
```diff
@@ -0,0 +1,49 @@
+                        requestModel.getInventory()
+                );
+
+                UseItemResponseModel responseModel = new UseItemResponseModel(1,
```
This implementation is valid only for one-time use items.

adimaggio2021 reviewed (dismissed):
Thorough code, your classes follow the clean architecture very well. Please see comments and watch with capitalization. The package should be lower-case mariogatchi and final variables should be all caps.

SpecialThing44 reviewed (dismissed):
I think I largely agree with Anthony, not much else to mention other than more commenting would be appreciated. Clean implementation otherwise.

Samuel-Glasroth reviewed (commented):
AddItemRequestModel and AddItemResponseModel  and RemoveItemRequestMdoel have the same code that should not be

In AddItemPresenter class in the addItemToInv function you should not be using instanceof funtion

In Inventory class in the addItem function you have a variable using snake case instead of camel case

Grace199 committed:
Grace - Added enums for Environments, Item Effects, and Items.
       - Simplified Items superclass (removed abstract method eligibleMariogatchi(), removed abstract method updateInventory())
        - Added two implemented Items. One is non-reusable (Apple), and the other is reusable (Leash).
        - Added a check in RemoveItemRunner for reusable/default items. If the Item is Reusable, then it cannot be removed from the User’s Inventory.

adimaggio2021 commented on src/main/java/Mariogatchi/Inventory.java:
```diff
@@ -29,7 +29,9 @@ public boolean removeItem(Item item, int quantityToRemove) {
             int newQuantity = itemToQuantity.get(item) - quantityToRemove;
             itemToQuantity.replace(item, newQuantity);
-
+            if (newQuantity == 0) {
```
👍

adimaggio2021 reviewed (approved):
Looks good, thanks for addressing comments.

Samuel-Glasroth reviewed (changes requested):
You still have not fixed the redundant code and removed the instanceof function

SpecialThing44 reviewed (approved):
Looks better than before, good work

Samuel-Glasroth reviewed (approved):
looks good
